microservices-connector/config/authN-authZ/README.md

```diff
@@ -156,6 +171,123 @@ curl -X POST $accessUrl -d '{"text":"What is the revenue of Nike in 2023?","para
 
-# try with the valid token from mary. Shall get the correct response from LLM
-curl -X POST $accessUrl -d '{"text":"What is the revenue of Nike in 2023?","parameters":{"max_new_tokens":17, "do_sample": true}}' -sS -H "Authorization: Bearer $TOKENA" -H 'Content-Type: application/json'
+# try with the valid token from mary. Shall get the correct response from LLM with http code
+curl -X POST $accessUrl -d '{"text":"What is the revenue of Nike in 2023?","parameters":{"max_new_tokens":17, "do_sample": true}}' -sS -H "Authorization: Bearer $TOKENA" -H 'Content-Type: application/json' -w " %{http_code}\n"
+```
+
+## Perform authentication and authorization via oauth2-proxy and OIDC provider
+
+Another choice we have is using oauth2-proxy and OIDC providers. These two streamline authentication and authorization by handling user identity and access management. oauth2-Proxy acts as a gateway, integrating with OIDC providers to authenticate users and issue tokens. This setup ensures secure access to applications by validating user credentials and managing permissions, simplifying the implementation of robust security protocols across services.
+
+<img src="./docs/OPEA auth flow with oauth2-proxy.png" width="700" height="400">
+
+We are using a similar scenario here that only privileged users can access our chatQnA service and ask questions. In this case, user `mary` who has the role `user` can access the chatQnA pipeline. And user `bob` with the role `viewer` will not be able to access the service. Of course, the other users without valid token cannot access the service.
+
+**Installation and configuration of oauth2-proxy and OIDC Provider**
+
+Here we take Keycloak as the sample OIDC Provider to use in the example. Follow the steps to install and configure Keycloak.
+
+```bash
+cd $(pwd)
+kubectl apply -f $(pwd)/config/authN-authZ/keycloak_install.yaml
+
+# get the ip and port to access keycloak.
+export HOST_IP=$(kubectl config view --minify -o jsonpath='{.clusters[0].cluster.server}' | cut -d '/' -f3 | cut -d ':' -f1)
+export KEYCLOAK_PORT=$(kubectl get svc keycloak -o jsonpath='{.spec.ports[0].nodePort}')
+export KEYCLOAK_ADDR=${HOST_IP}:${KEYCLOAK_PORT}
+```
+
+**Note:** Double check if the host ip captured is the correct ip.
+
+Access the Keycloak admin console through the `KEYCLOAK_ADDR` to configure the users as well as roles. Use the username and password specified in the yaml file to login.
+
+The user management is done via Keycloak and the configuration steps look like this:
+
+1. Create a new realm named `chatqna` within Keycloak.
+
+<img src="./docs/create_realm.png" width="600" height="300">
+
+2. Create a new client called `chatqna` and set `Client authentication` to 'On'. Set "http://chatqna-ui.com:${INGRESS_PORT}/*" in the `Valid redirect URIs` part. Under the Credentials tab you will now be able to locate `<your client's secret>`, which will be used in the oauth2-proxy configs.
```

Comment:
Hi @Ruoyu-y, we may need to tell user how to obtain this `INGRESS_PORT`,  by executing shell command or a link of Istio?

Reply:
The same to `INGRESS_HOST`

Reply:
Done